site/blog/2025-07-22-ha-postgres-full-sync.mdx

```diff
@@ -0,0 +1,84 @@
+    * the primary is restarted after a failover, it will have a timeline that is incompatible with the rest of the cluster, and will not be able to rejoin the cluster.
+    * the client that initiated the transaction breaks the connection, the transaction becomes visible, and can be read by other connections. After this read, the node can fail before the WAL replicates, resulting in phantom reads. Although this is rare, it’s theoretically possible.
+    * replica that is not a hot standby may receive a transaction before the standby. This becomes a problem if the failover process does not account for this possibility.
+2. The primary externalizes a transaction only after receiving an ack from the replica. But the replica itself will externalize it as soon as it’s received it: An application may see this committed transaction on the replica, but may not be able to find it on the primary.
+    
+    <img src="/img/blog/fsync2.svg" alt="fsync2" />
+    
+3. Failures during failover: If there are multiple failures during a failover, we may end up with multiple conflicting timelines. There is no methodical way to determine which one is authoritative.
+
+    <img src="/img/blog/fsync3.svg" alt="fsync3" />
+
+The above failure modes get progressively more unwieldy as the number of nodes increases.
+
+## Multigres solution
+
+Some of these problems have mitigations, but not all are solvable. With Multigres, we plan to address some of these issues as follows:
+
+- **1a: CDC picking up a transaction early:** Run logical replication on a replica.
+- **1b: Primary crashes after writing a transaction to the WAL:** Multigres can automate the repair of the primary when it comes back up.
+- **1c: Client reading phantom records:** There is no mitigation, but this is a rare occurrence.
+- **1d: Replica that is not a standby receives errant transaction:** Multigres can automate the repair.
+- **2: Replica committing before primary:** There is no mitigation, but the application can work around this possibility on a case-by-case basis.
+- **3: Failures during failover:** Assuming clocks are reliable, Multigres can use the WAL timestamps to determine which of the failover attempts was the latest and can use it. We will cover why this is the most authoritative timeline when we explain consensus algorithms in upcoming blogs.
+
+Many of these problems can be tolerated by the application, but it will be nice to eventually eliminate them altogether. This is why we intend to implement a two-phase sync replication solution, which can be the foundation for consensus protocols.
+
+Furthermore, the two-phase approach will have the same performance overhead as the full-sync approach, so you have nothing to lose and everything to gain by transitioning to it. In any case, until it becomes robust, the full-sync approach can continue be used as an acceptable compromise.
```

Comment:
that's an interesting statement! I understand that you feel it, and I'm still curious

I'm still learning and would be great to do some benchmarks / source code research to see how they compare performance-wise